integrate/simpsons.go

```diff
@@ -0,0 +1,96 @@
+// Copyright ©2019 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import (
+	"sort"
+)
+
+// Simpson's method approximates the integral of a function f(x)
+// by means of subdividing the interval of integration into segments
+// and applying a method that fits a polynomial to each subinterval.
+// this implementation makes use of the following composite Simpson's method
+// to estimate \int_a^b f(x) dx:
+//  \sum_{i=0}^{(N/2)-1} {a_0}_i * f_{2i-1} + {a_1}_i * f_{2i} + {a_2}_i * f_{2i+1}
+// where N is the number of subintervals and the number of subintervals
+// is even and {a_0}_i, {a_1}_i, and {a_2}_i are constants at index i given by:
+//  {a_0}_i = 2 * h^{3}_{2i} - h^{3}_{2i+1} + 3 * h_{2i+1} * h^{2}_{2i} /
+//            6 * h_{2i} * (h_{2i+1} + h_{2i})
+//  {a_1}_i = h^{3}_{2i+1} + h^{3}_{2i} +
+//            3 * h_{2i+1} * (h_{2i+1} + h_{2i}) /
+//            6 * h_{2i+1} * h_{2i}
+//  {a_2}_i = 2 * h^{3}_{2i+1} - h^{3}_{2i} + 3 * h_{2i} * h^{2}_{2i+1} /
+//            6 * h_{2i+1} * (h_{2i+1} + h_{2i})
+// where h_{k} is the length of the subinterval at k.
+//
+// The formula above approximates the integral of function of f
+// if N is an even number of subintervals. If the number of subintervals are odd,
+// the subintervals up i=0..n-2 are given by the above
+// and the approximations over the second to last and last subintervals are given by:
+//  {a_0}_i * f_{N-2} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N}
+// where the coefficients are:
+//  {a_0}_i = -1 * h^{3}_{N-1} /
+//             6 * h_{N-2} * (h_{N-2} + h_{N-1})
+//  {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
+//            6 * h_{N-2}
+//  {a_2}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
+//            6 * (h_{N-2} + h_{N-1})
+// More information is available at:
+// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
+//
+// The (x,f) input data points must be sorted along x. One can use stat.SortWeighted to do that.
+// The x and f slices must be of equal length and have length > 1.
```

Comment:
s/> 1/greater than one/

Reply:
greater than two

Reply:
Indeed.

Reply:
Oops. I made the following changes? What additional changes are needed?

- Updated length
- Reduced spacing
- Defined a,b
- Defined h_{k}
- matched the description of the first formula to the implementation `\sum_{i=1}^{N} {a_0}_i * f_{2i} + {a_1}_i * f_{i} + {a_2}_i * f_{i+1}`

```
// Simpson's method approximates the integral of a function f(x)
// by means of subdividing the interval of integration into segments
// and applying a method that fits a polynomial to each subinterval.
// This implementation makes use of the following composite Simpson's method
// to estimate \int_a^b f(x) dx where a, b are x at x[0] and x[len(x)-1] respectively:
//  \sum_{i=1}^{N} {a_0}_i * f_{2i} + {a_1}_i * f_{i} + {a_2}_i * f_{i+1}
// where N is the count of subintervals and {a_0}_i, {a_1}_i, and {a_2}_i are constants at index i given by:
//  {a_0}_i * f_{N-2} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N}
// where the coefficients are:
//  {a_0}_i = 2 * h^{3}_{i} - h^{3}_{i+1} + 3 * h_{i+1} * h^{2}_{i} /
//            6 * h_{i} * (h_{i} + h_{i+1})
//  {a_1}_i = h^{3}_{i} + h^{3}_{i+1} +
//            3 * h_{i} * h_{i+1} * (h_{i} + h_{i+1}) /
//            6 * h_{i} * h_{i+1}
//  {a_2}_i = -h^{3}_{i} + 2 * h^{3}_{i+1} + 3 * h_{i} * h^{2}_{i+1} /
//            6 * h_{i+1} * (h_{i} + h_{i+1})
// where h_{k} is the difference x[k] - x[(k-1)].
// The formula above approximates the integral of function of f if N is an even number of subintervals.
// If the number of subintervals are odd, the subintervals up i=0..n-2 are given by the above
// and the approximations over the second to last and last subintervals are given by:
//  {a_0}_i * f_{N-2} + {a_1}_i * f_{N-1} + {a_2}_i * f_{N}
// where the coefficients are:
//  {a_0}_i = -1 * h^{3}_{N-1} /
//             6 * h_{N-2} * (h_{N-2} + h_{N-1})
//  {a_1}_i = h^{2}_{N-1} + 3 * h^{3}_{N-1} * h_{N-2} /
//            6 * h_{N-2}
//  {a_2}_i = 2 * h^{2}_{N-1} + 3 * h^{2}_{N-1} * h_{N-2} /
//            6 * (h_{N-2} + h_{N-1})
// More information is available at:
// https://en.wikipedia.org/wiki/Simpson%27s_rule#Composite_Simpson's_rule_for_irregularly_spaced_data
//
// The (x,f) input data points must be sorted along x. One can use stat.SortWeighted to do that.
// The x and f slices must be of equal length and the length must be greater than two.
```